arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+          std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(f._lastRequestLatency)
+              .count()};
+    }
+    return LogStatus{std::move(status)};
+  });
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, bool waitForSync) -> LogIndex {
+  auto index = insert(std::move(payload), waitForSync, doNotTriggerAsyncReplication);
+  triggerAsyncReplication();
+  return index;
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, [[maybe_unused]] bool waitForSync,
+                                       DoNotTriggerAsyncReplication) -> LogIndex {
+  // TODO Handle waitForSync!
+  auto const insertTp = InMemoryLogEntry::clock::now();
+  // Currently we use a mutex. Is this the only valid semantic?
+  return _guardedLeaderData.doUnderLock([&](GuardedLeaderData& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    auto const index = leaderData._inMemoryLog.getNextIndex();
+    auto const payloadSize = payload.byteSize();
+    auto logEntry =
+        InMemoryLogEntry(PersistingLogEntry(_currentTerm, index, std::move(payload)));
+    logEntry.setInsertTp(insertTp);
+    leaderData._inMemoryLog.appendInPlace(_logContext, std::move(logEntry));
+    _logMetrics->replicatedLogInsertsBytes->count(payloadSize);
+    return index;
+  });
+}
+
+auto replicated_log::LogLeader::waitFor(LogIndex index) -> WaitForFuture {
+  return _guardedLeaderData.doUnderLock([index](auto& leaderData) {
+    if (leaderData._didResign) {
+      auto promise = WaitForPromise{};
+      promise.setException(basics::Exception(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED,
+                                             __FILE__, __LINE__));
+      return promise.getFuture();
+    }
+    if (leaderData._commitIndex >= index) {
+      return futures::Future<std::shared_ptr<QuorumData const>>{std::in_place,
+                                                                leaderData._lastQuorum};
+    }
+    auto it = leaderData._waitForQueue.emplace(index, WaitForPromise{});
+    auto& promise = it->second;
+    auto&& future = promise.getFuture();
+    TRI_ASSERT(future.valid());
+    return std::move(future);
+  });
+}
+
+auto replicated_log::LogLeader::getParticipantId() const noexcept -> ParticipantId const& {
+  return _id;
+}
+
+auto replicated_log::LogLeader::triggerAsyncReplication() -> void {
+  auto preparedRequests = _guardedLeaderData.doUnderLock([](auto& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    return leaderData.prepareAppendEntries();
+  });
+  executeAppendEntriesRequests(std::move(preparedRequests), _logMetrics);
+}
+
+auto replicated_log::LogLeader::GuardedLeaderData::updateCommitIndexLeader(
+    LogIndex newIndex, std::shared_ptr<QuorumData> const& quorum) -> ResolvedPromiseSet {
```

Comment:
It looks like we could receive `quote` by value here, so that we could move it into the `ResolvedPromiseSet` later
```suggestion
    LogIndex newIndex, std::shared_ptr<QuorumData> quorum) -> ResolvedPromiseSet {
```